### PR TITLE
feat: improved output format

### PR DESCRIPTION
**Pull request by RomainMuller**
Improved the TypeScript output by providing a new `CodeSink` output API. This allows managing indentation in a producer-independent manner. This allows properly indenting the generated code in a simple and concise manner.

In addition to this, some TypeScript style changes include:
- `CfnOutput`s are only created if they have an `Export` property
- All `CfnOutput`s result in a `readonly` property being visible on the Stack
- Optional resources are defined in `const` bindings thanks to the `?:` operator
- Using `CfnResource#cfnOptions` & `CfnResource#addDependency` properties instead of using `.addOverride`
- `Fn::Base64` now decodes in-line if the string is "hard-coded" & contains valid UTF-8
- `Fn::Select` extracts the right element if the list is hard-coded

This also comes with a significant refactor of the TypeScript emitter in smaller functions that are easier to process independently.

### CODE REVIEW

**Review by iph — APPROVED** [review]
Lgtm, I need to think about the ramifications of nuking empty outputs on correctness (e.g. I use outputs in terraform to link things) but overall love the changes!

File: src/synthesizer/output.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        indentation.push_str(self.config.indent);
+        CodeSink {
+            config: self.config,
+            indentation,
+            writer: self.writer,
+        }
+    }
+
+    /// Writes a blank line to the underlying writer with no indentation.
+    pub fn blank_line(&mut self) -> io::Result<()> {
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply src/synthesizer/output.rs:36]: Yes... Think for example, a `Vec<u8>` is a `std::io::Write` implementor... And so you can't add bytes into it unless you have mutable access to it... It also saves you from data races from concurrent writes :)

File: src/synthesizer/typescript_synthesizer.rs [hunk 2]
```diff
@@ -454,12 +514,38 @@ pub fn to_string_ir(resource_value: &ResourceIr) -> Option<String> {
             let count_str = to_string_ir(count.as_ref()).unwrap();
             let cidr_bits_str = to_string_ir(cidr_bits.as_ref()).unwrap();
             Option::Some(format!(
-                "cdk.Fn.cidr({ip_block_str}, {count_str}, {cidr_bits_str})"
+                "cdk.Fn.cidr({ip_block_str}, {count_str}, '{cidr_bits_str}')"
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply src/synthesizer/typescript_synthesizer.rs:517]: Yeah I'm mega surprised... likely will have to research this a bit further...

**Comment by RomainMuller** [comment]
> Lgtm, I need to think about the ramifications of nuking empty outputs on correctness (e.g. I use outputs in terraform to link things) but overall love the changes!

We can absolutely add an option such as `--retain-all-cfn-outputs` or something... which makes the emit of the `CfnOutput` unconditional...

**Pull request merged**